spec/did.md

```diff
@@ -0,0 +1,79 @@
+| `keyAgreement`       | Array of strings | No   | String values must be fully qualified DID URIs (e.g. `did:example:abcd#key-1` over `#key-1`). |
+| `capabilityInvocation` | Array of strings | No | String values must be fully qualified DID URIs (e.g. `did:example:abcd#key-1` over `#key-1`). |
+| `capabilityDelegation` | Array of strings | No | String values must be fully qualified DID URIs (e.g. `did:example:abcd#key-1` over `#key-1`). |
+| `service`            | Array of [Services](#service-data-model) | No | - |
+
+**Additional Notes:**
+- No [JSON-LD processing](https://www.w3.org/TR/did-core/#consumption-0) is performed.
+- Each [Verification Method](https://www.w3.org/TR/did-core/#verification-methods) must have at least one [Verification Relationship](https://www.w3.org/TR/did-core/#verification-relationships).
+
+### Verification Method Data Model
+
+Following form [this data model](https://www.w3.org/TR/did-core/#verification-methods).
+
+| Property      | JSON Representation | Required | Notes          |
+| ------------- | ------------------- | -------- | -------------- |
+| `id`          | String              | Yes      | Must be a fully qualified DID URI (e.g. `did:example:abcd#key-1`). |
+| `type`        | String              | Yes      | Must be a URI. |
+| `controller`  | String              | Yes      | Must be a URI. |
+| `publicKeyJwk`        | Object      | No       | Preferred. Represents a [JWK](https://www.w3.org/TR/did-core/#bib-rfc7517). Either this or `publicKeyMultibase` must be present. |
+| `publicKeyMultibase`  | String      | No       | Not preferred. Either this or `publicKeyJwk` must be present. |
+
+### Service Data Model
+
+Following from [this data model](https://www.w3.org/TR/did-core/#services).
+
+| Property          | JSON Representation | Required | Notes          |
+| ----------------- | ------------------- | -------- | -------------- |
+| `id`              | String              | Yes      | Must be a fully qualified DID URI (e.g. `did:example:abcd#service-1`). |
+| `type`            | String              | Yes      | Must be a type defined in the [service registry](https://www.w3.org/TR/did-spec-registries/#service-types). |
+| `serviceEndpoint` | Array of Strings    | Yes      | String values must be URIs. |
+| `sig`             | Array of Strings    | No       | -                           |
+| `enc`             | Array of Strings    | No       | -                           |
+
+## DID Resolution Metadata Data Model
+
+DID Resolution Metadata is _always optional_.
+
+Following from [this data model](https://www.w3.org/TR/did-core/#did-resolution-metadata).
+
+| Property      | JSON Representation | Required | Notes          |
+| ------------- | ------------------- | -------- | -------------- |
+| `error`       |  String             | No       | Required if there was an error during resolution. One of [`invalidDid`, `notFound`, `representationNotSupported`]. |
```

Comment:
thoughts on including [`methodNotSupported`](https://www.w3.org/TR/did-spec-registries/#methodnotsupported)?

Reply:
we also have `internalError` as one of the error enums

Reply:
➕ 1 on adding `methodNotSupported` and `internalError`.

I also think we ought to consider registering additional errors in the [DID Resolution Spec](https://w3c-ccg.github.io/did-resolution/) that are particularly relevant for DID DHT resolution:
- `invalidDidDocument`: The DID document supplied does not conform to valid syntax.
- `invalidDidDocumentLength`: The byte length of a DID document is not within an acceptable range, similar to the Web IDL / DOM [RangeError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RangeError)

Reply:
will add both

what should `internalError` mean?

Reply:
@decentralgabe Missed this question.  I've always interpreted `internalError` similar to the role HTTP Status code 500 plays.  The (draft) [spec text](https://w3c-ccg.github.io/did-resolution/#internalerror) reads:
> When an unexpected error occurs during [DID Resolution](https://w3c-ccg.github.io/did-resolution/#dfn-did-resolution) or [DID URL dereferencing](https://w3c-ccg.github.io/did-resolution/#dfn-did-url-dereferencing), the value of the DID Resolution or DID URL Dereferencing Metadata error property MUST be internalError.

Given that the error is "unexpected" it covers cases of errors that aren't mapped to a more specific error code.